File: 0889-construct-binary-tree-from-preorder-and-postorder-traversal/0889-construct-binary-tree-from-preorder-and-postorder-traversal.java
class TreeNode {
    int val;
    TreeNode left;
    TreeNode right;

    TreeNode(int x) {
        val = x;
    }
}

class Solution {
    public TreeNode constructFromPrePost(int[] preorder, int[] postorder) {
        HashMap<Integer, Integer> postIndexMap = new HashMap<>();
        for (int i = 0; i < postorder.length; i++) {
            postIndexMap.put(postorder[i], i);
        }
        return buildTree(preorder, postIndexMap, 0, preorder.length - 1, 0);
    }

    private TreeNode buildTree(int[] preorder, HashMap<Integer, Integer> postIndexMap, 
                               int preStart, int preEnd, int postStart) {
        if (preStart > preEnd) {
            return null; 
        }
        
        TreeNode root = new TreeNode(preorder[preStart]);
        
        if (preStart == preEnd) {
            return root;
        }
        
        int leftChild = preorder[preStart + 1];
        int leftChildIndex = postIndexMap.get(leftChild);
        int leftSubtreeSize = leftChildIndex - postStart + 1;
        root.left = buildTree(preorder, postIndexMap, preStart + 1, preStart + leftSubtreeSize, postStart);
        root.right = buildTree(preorder, postIndexMap, preStart + leftSubtreeSize + 1, preEnd, leftChildIndex + 1);
        return root;
    }
}

File: 0889-construct-binary-tree-from-preorder-and-postorder-traversal/0889-construct-binary-tree-from-preorder-and-postorder-traversal.py
class Solution:
    def constructFromPrePost(self, preorder: List[int], postorder: List[int]) -> Optional[TreeNode]:
        post_map = {val: idx for idx, val in enumerate(postorder)}
        
        def build(pre_start, pre_end, post_start, post_end):
            if pre_start > pre_end:
                return None
                
            root = TreeNode(preorder[pre_start])
            
            if pre_start == pre_end:
                return root
                
            left_val = preorder[pre_start + 1]
            left_subtree_end = post_map[left_val]
            left_subtree_size = left_subtree_end - post_start + 1
            
            root.left = build(pre_start + 1, 
                            pre_start + left_subtree_size,
                            post_start,
                            left_subtree_end)
            
            root.right = build(pre_start + left_subtree_size + 1,
                             pre_end,
                             left_subtree_end + 1,
                             post_end - 1)
            
            return root
        
        return build(0, len(preorder) - 1, 0, len(postorder) - 1)